### ISSUE TITLE
Request to create firmware for space mouse classic usb

### ISSUE THREAD

**Issue by Platidors** [jfedor2/magellan-spacemouse#1]
Hello. I assembled [hid-remapper](https://github.com/jfedor2/hid-remapper) for your project on one raspberry pi pico board. Tried connecting my old Space Mouse Classic USB controller. On some computers, the Space Mouse Classic USB is not installed correctly. I tried with the help of your device to solve this problem. Space Mouse certainly didn't work with [hid-remapper](https://github.com/jfedor2/hid-remapper). I want to ask you to make Space Mouse Compact emulation firmware for Space Mouse Classic USB based on the [hid-remapper](https://github.com/jfedor2/hid-remapper) board. Ready to become a firmware tester.

![photo_2023-03-08_22-54-15](https://user-images.githubusercontent.com/127308740/223808116-c5f107d0-e503-42a2-9506-cfcbf60217ff.jpg)

![photo_2023-03-08_22-56-41](https://user-images.githubusercontent.com/127308740/223808699-2caf9f47-438f-4d35-8f72-474cef2f456e.jpg)

![photo_2023-03-08_22-56-51](https://user-images.githubusercontent.com/127308740/223808705-6dfa91a5-60be-4aea-a820-d5e177c9e3f3.jpg)

**Comment by jfedor2** [comment]
Do you have means to dump the HID report descriptor of the device?

On Linux it can be done with usbhid-dump or hid-recorder from the hid-tools package. And on Windows with Wireshark.

Even now you should probably be able to map the device's buttons/axes to mouse or keyboard with HID Remapper, it's just that you have to know the HID usage codes that it sends. Does is show some strange looking hex codes in the input list when it's connected? Like 0x00010033 or something.

**Comment by jfedor2** [comment]
Nice enclosure BTW.

**Comment by Platidors** [comment]
I tried using Wireshark. I don't know if I did it right.

[usb.zip](https://github.com/jfedor2/magellan-spacemouse/files/10925003/usbЯ

**Comment by Platidors** [comment]
[usb.zip](https://github.com/jfedor2/magellan-spacemouse/files/10925019/usb.zip)
I tried using Wireshark. I don't know if I did it right.

**Comment by jfedor2** [comment]
I can see the device and configuration descriptors (and actual traffic), but not the report descriptor. I think there was some trick to getting it with Wireshark. Did you unplug and re-plug the device while capturing?

**Comment by Platidors** [comment]
[USB2.zip](https://github.com/jfedor2/magellan-spacemouse/files/10925203/USB2.zip)
here is the disconnect and connect file

**Comment by jfedor2** [comment]
Yeah, that's it!

So based on this, assuming HID Remapper works with this device on hardware level, when you connect it and just keep the default configuration, the button labelled "1" should work as the left mouse button. If that's working then we can try to make it emulate a modern SpaceMouse so that it works with current software.

**Comment by Platidors** [comment]
It didn't work for me. Or am I doing something wrong. When a regular mouse is connected, the LED blinks, when a space mouse is connected, the LED does not glow. Button "1" does nothing.

**Comment by jfedor2** [comment]
The single Pico version doesn't work with some devices, even if such cases are rare with the newest version. You could try making the dual Pico variant.

**Comment by Platidors** [comment]
The dual version HID-remapper will still not work as an emulator. What can be obtained by making a double version?

**Comment by jfedor2** [comment]
We can easily make it into an emulator by changing the output report descriptor (and the USB vendor/product IDs). But it has to be able to receive inputs from your Classic, which the single Pico version seems to have trouble with.

**Comment by Platidors** [comment]
Okay, I'll try to build a double version tomorrow

**Comment by Platidors** [comment]
Today I collected a double version. When you connect a space mouse, it starts acting on the cursor. When the puck moves, the cursor moves to the lower right corner, pressing the buttons has the same effect on the cursor. The LED on  the remapper starts blinking when the space mouse is actuated.

![photo_2023-03-10_16-46-47](https://user-images.githubusercontent.com/127308740/224321134-874cdc5b-fb8a-49da-922b-d5e1ad26632f.jpg)

![photo_2023-03-10_16-46-42](https://user-images.githubusercontent.com/127308740/224321137-c8a7203e-ddf7-4da3-8003-e8a4a862998e.jpg)

**Comment by jfedor2** [comment]
OK, this is a good sign I think.

Try flashing your HID Remapper with this test firmware:

[remapper_dual_a_spacemouse.zip](https://github.com/jfedor2/magellan-spacemouse/files/10942414/remapper_dual_a_spacemouse.zip)

With this firmware it should pretend to be a SpaceMouse Compact. Because of this you won't be able to use the web configuration tool to change the mappings, but the default passthrough configuration should be good at least to test.

When I tested with a real SpaceMouse Compact connected to it, I noticed some choppiness in the movements so there's probably more to look into, but it worked well enough to fool the 3Dconnexion drivers and Fusion 360.

**Comment by Platidors** [comment]
After the firmware, the remapper became defined as a compact space mouse. Pressing the buttons and moving the joystick opens the radial menu 3Dconnexion. The object in the programs does not move

**Comment by jfedor2** [comment]
Weird. It's as if we're somehow misinterpreting the data coming from the device, but I didn't notice any strange things in the descriptor or the captured traffic that could be the cause. Let me think of some way to debug this.

**Comment by jfedor2** [comment]
Do you have a USB-TTL adapter by any chance? To read serial output from the Pico?

**Comment by Platidors** [comment]
Unfortunately, no such device is available.

**Comment by Platidors** [comment]
It was necessary to flash only part A? When idle, the reassigner constantly sends a report request
![2023-03-10 193945](https://user-images.githubusercontent.com/127308740/224359810-e66fb234-46fb-4d7b-8983-33070e96b40a.png)

**Comment by jfedor2** [comment]
What's inside these reports (responses in particular)? It might be that there's some noise in the sensors and your device keeps sending reports as well and the remapper passes them on to the PC.

And let me make sure I understood correctly, does the 3Dconnexion radial menu open even when not pressing any buttons on the device (just moving the controller)?

**Comment by Platidors** [comment]
Yes, any movement of the space mouse classic works out pressing 2 buttons of the space mouse compact

**Comment by Platidors** [comment]
The report request is constantly even if you disconnect the space mouse from the remapper
[USB3.zip](https://github.com/jfedor2/magellan-spacemouse/files/10944044/USB3.zip)

**Comment by jfedor2** [comment]
This in itself is not a problem I think. It happens because in the SpaceMouse Classic report descriptor, all the X/Y/Z/RX/RY/RZ inputs are defined as absolute. In the SpaceMouse Compact report descriptor, which we're using for the PC side, they are defined as relative (which is a mistake in my opinion, but it doesn't matter normally). HID Remapper just repeats the last seen state, even if the device that originally generated that state was disconnected.

Maybe I will try and program another HID Remapper to emulate a SpaceMouse Classic, connect it to the first one and see if there's something in the behavior that I missed.

**Comment by jfedor2** [comment]
OK, I've found one issue that was preventing the axes from working properly.

This doesn't explain the button behavior, but at least maybe the rotations/translation will work now.

Try this firmware.

[remapper_dual_a_spacemouse2.zip](https://github.com/jfedor2/magellan-spacemouse/files/10944867/remapper_dual_a_spacemouse2.zip)

**Comment by Platidors** [comment]
The behavior of the joystick on this firmware has not changed. The command to launch the two buttons of the space mouse compact

**Comment by jfedor2** [comment]
Hmm then I'm out of ideas for now.

**Comment by Platidors** [comment]
I have an arduino nano. I can make a ttl usb converter out of it.

**Comment by jfedor2** [comment]
Sounds like a plan. The default TX pin on the dual HID Remapper is GPIO 0 and the baud rate is 921600 (connect it to the A side Pico). By default it just prints the number of received and sent reports every second. Let me know if you can see that and I'll make a build with some diagnostics.

**Comment by Platidors** [comment]
Conected
![image](https://user-images.githubusercontent.com/127308740/224485121-11b85d95-c9d6-4d65-82df-196c86e574ed.png)

**Comment by jfedor2** [comment]
It should currently be outputting something like this (in readable ASCII), assuming no inputs are sent:

```
0 0
0 0
0 0
```

Not sure what the setup is, but you probably need to set the parameters in the Arduino sketch, not on PC side (baud rate etc.).

**Comment by jfedor2** [comment]
Data bits should be 8 (again, on both sides).

**Comment by Platidors** [comment]
![image](https://user-images.githubusercontent.com/127308740/224486200-2180cd82-e318-43da-9da4-69629e9457a1.png)

**Comment by jfedor2** [comment]
Great!

Try this then.

[remapper_dual_a_spacemouse3.zip](https://github.com/jfedor2/magellan-spacemouse/files/10948639/remapper_dual_a_spacemouse3.zip)

It should output something like this when moving the joystick:

```
recv: 00 00 a1 ff 00 00 15 00 00 00 00 00 00 00 
00010035: 0
00010034: 0
00010033: 0
00090001: 0
00090002: 0
00090003: 0
00090004: 0
00090005: 0
00090006: 0
00090007: 0
00090008: 0
00090009: 0
0009000c: 0
00010031: 0
00010030: -95
00010032: 21
```

It's what it receives and how it interprets it. 0009000x are buttons, 0001003x are translations/rotations. In the "recv" line, there should be two bytes for the buttons and then two bytes for X, Y, Z, RX, RY, RZ. Try to see if it corresponds to what you're doing with the controller. Particularly if it looks like the buttons are activated when they're not.

You might also get some "overflow!" messages, don't worry about them for now, I will think on how to fix it, but it should be a separate issue.

**Comment by Platidors** [comment]
[joystick.txt](https://github.com/jfedor2/magellan-spacemouse/files/10948663/joystick.txt)
[key1.txt](https://github.com/jfedor2/magellan-spacemouse/files/10948664/key1.txt)

**Comment by jfedor2** [comment]
Weird! It looks like we're receiving the reports split into two parts, 8+6 bytes. Not sure why.

(EDIT: I guess it's split because the maximum packet length is 8 and the report length is 14. Should this be handled by the application on the receiving end? Or somewhere in the library?)

Anyway, try this build with a workaround:

[remapper_dual_a_spacemouse4.zip](https://github.com/jfedor2/magellan-spacemouse/files/10948710/remapper_dual_a_spacemouse4.zip)

**Comment by Platidors** [comment]
Yes, it works. Buttons 1 and 2 work as configured in 3Dconexion control center . The joystick also works on all axes, but in my opinion some axis does not work there. But that's good.)

**Comment by Platidors** [comment]
Confused "up" "down" with "zoom in" "zoom out"

**Comment by jfedor2** [comment]
Cool. Here's a build with the diagnostics removed. Let me know if you're seeing any "overflow!" messages while using the joystick.

[remapper_dual_a_spacemouse5.zip](https://github.com/jfedor2/magellan-spacemouse/files/10948799/remapper_dual_a_spacemouse5.zip)

**Comment by Platidors** [comment]
[overflow.txt](https://github.com/jfedor2/magellan-spacemouse/files/10948884/overflow.txt)

**Comment by jfedor2** [comment]
OK, this should solve that issue.

[remapper_dual_a_spacemouse6.zip](https://github.com/jfedor2/magellan-spacemouse/files/10949161/remapper_dual_a_spacemouse6.zip)

Do the axes work correctly now, by the way? You can configure a lot of the behavior in 3Dconnexion's utility.

**Comment by Platidors** [comment]
Yes, everything seems to be fine in the terminal. 

![image](https://user-images.githubusercontent.com/127308740/224506723-fab2ba96-5d92-4f95-8cb6-951a2c2da0ce.png)

In software 3Dconnexion control panel, I can only invert the direction of the axes. But I can't replace one axle with another.  Axis "up" "down" should be replaced with "zoom in" "zoom out". When the joystick is raised, the object is approaching, it must go up

**Comment by jfedor2** [comment]
Actually there is a "Zoom Direction" setting, do you have it set to "Forward / Backward"?

Do the rotation axes work the way you expect them to?

Swapping axes is of course not a problem, it is after all a HID Remapper.

**Comment by Platidors** [comment]
rotation and tilt seem to work correctly

**Comment by Platidors** [comment]
> Actually there is a "Zoom Direction" setting, do you have it set to "Forward / Backward"?

I tried switching, it gets worse

**Comment by jfedor2** [comment]
OK, this swaps Y and Z axes, which I think is what we want.

[remapper_dual_a_spacemouse7.zip](https://github.com/jfedor2/magellan-spacemouse/files/10949322/remapper_dual_a_spacemouse7.zip)

**Comment by Platidors** [comment]
The axes are swapped, the zoom works well. up and down works inversely. I led you into confusion, the tilt and turn axes are the same mixed up.

**Comment by jfedor2** [comment]
OK, so like I said, since this is actually HID Remapper, you can swap and invert the axes yourself. The only problem is that since we're faking the USB Vendor/Product IDs, the regular web config interface won't recognize the device. So I made a custom version that ignores the IDs and lets you select from the list of all devices.

Load this firmware (the previous one has mappings disabled):

[remapper_dual_a_spacemouse8.zip](https://github.com/jfedor2/magellan-spacemouse/files/10949483/remapper_dual_a_spacemouse8.zip)

And go here to set up the mappings:

https://www.jfedor.org/misc2/spacemouse-remap/

For example like this:
![Screenshot from 2023-03-11 22-15-47](https://user-images.githubusercontent.com/2271721/224512048-f6272479-7ae7-4438-a158-83771cc201f5.png)
Or you might need to set scaling to -1 for one of the rotation axes as well. Play around. If you haven't used HID Remapper before there's some documentation in the repository README.

**Comment by Platidors** [comment]
Thanks a lot for the great work. Finally managed to adjust the axis. The configuration looks like this
![image](https://user-images.githubusercontent.com/127308740/224513191-fc71f6eb-054b-433e-8276-e07d934eccac.png)

[hid-remapper-config.zip](https://github.com/jfedor2/magellan-spacemouse/files/10949537/hid-remapper-config.zip)

And another question. Is it possible to assign any buttons or keyboard shortcuts to the remaining buttons?

**Comment by jfedor2** [comment]
Not with the current version as I replaced the keyboard and mouse stuff with spacemouse emulation, but maybe it could be added back on a separate interface or something.

**Comment by Platidors** [comment]
I have a space explorer, it has 16 buttons. You can try to emulate it. Although it has been removed from support, it still works properly in all programs with 10.6.4 drivers. If you want to let me know, you can experiment with it. But even now it works very well.

**Comment by jfedor2** [comment]
If you can, just dump the report descriptor with Wireshark like you did with the Classic.

Ideally I'd like to get the descriptor for one of the modern supported devices with multiple buttons (SpaceMouse Pro or Enterprise).

**Comment by Platidors** [comment]
I'll try to find Space Mouse Pro and ask them to send a descriptor report

**Comment by Platidors** [comment]
I noticed that pressing the buttons from 1-8 causes a slight movement of the axis. This is not a big problem, but in sketch mode it can get in the way a bit. Can this be fixed somehow?

**Comment by jfedor2** [comment]
I'm guessing this is a physical thing, pressing the buttons causes some tiny readings on the sensors because the case flexes a bit or something like that.

You can try setting a dead zone, I believe on this model it can be done on device side by pressing "*"+"7" buttons. You can also set a dead zone in HID Remapper using expressions, that is a bit more involved, but nothing crazy. There are examples in the docs.

**Comment by Platidors** [comment]
"*"+"7" combination fails to set zero radius. It seems to me that this bug is not of a physical nature.

**Comment by Platidors** [comment]
![image](https://user-images.githubusercontent.com/127308740/224639252-cb0aad2c-9b70-478a-8508-78f66b280117.png)
Maybe the buttons at the bottom with addresses have something to do with it?

**Comment by Platidors** [comment]
Found a programming manual. Maybe you've already seen it, but still
https://www.manualslib.com/manual/1994332/Logitech-Magellan-Space-Mouse-Classic.html#manual

**Comment by Platidors** [comment]
Space Mouse Pro

**Comment by Platidors** [comment]
Space Mouse Pro
[SpaceMouseProUSBHook.zip](https://github.com/jfedor2/magellan-spacemouse/files/10955798/SpaceMouseProUSBHook.zip)

**Comment by jfedor2** [comment]
Thank you!

0x00090009 and 0x0009000c are buttons 9 and 12, they're in the descriptor, but the device doesn't even have that many. I guess maybe "*" is one of them, but I don't know if it's even sent to the host as it is used for device-side configuration (easy to test).

**Comment by Platidors** [comment]
May I hope for an emulator firmware based Space Mouse Pro?

**Comment by jfedor2** [comment]
Yeah, try this!

[remapper_spacemouse9.zip](https://github.com/jfedor2/magellan-spacemouse/files/10979547/remapper_spacemouse9.zip)

It also includes a different fix for the split reports issue so you need to flash both A and B sides. Let me know if there's any issue.

The buttons will need to be mapped.

The Classic has the following buttons:

1, 2, 3, 4, 5, 6, 7, 8, 9, 12

The Pro has the following:

1, 2, 3, 5, 6, 9, 13, 14, 15, 16, 23, 24, 25, 26, 27

**Comment by Platidors** [comment]
There were a lot of buttons with obscure addresses

![image](https://user-images.githubusercontent.com/127308740/225377220-b616adba-b453-4d62-a9b1-f71a53fbf222.png)

**Comment by Platidors** [comment]
Buttons with addresses do nothing. The rest of the buttons are reassigned normally. The axes seem to work correctly, but you have to swap them as before. Pressing the buttons still slightly move the joystick in 3D view.

**Comment by Platidors** [comment]
Button functions
Space Mouse Pro
1 - MENU
2 - Fit
3 - TopV
5 - RightV
6 - FrontV
9 - Roll+
13 - 1
14 - 2
15 - 3
16 - 4
23 - Esc
24 - Alt
25 - Shift
26 - Ctrl
27 - LockR 

Space Mouse Classic USB
1 - 1
2 - 2
3 - 3
4 - 4
5 - 5
6 - 6
7 - 7
8 - 8
9 - *

**Comment by Platidors** [comment]
Different behavior in the test program 3DxTest.exe.
Coordinates don't reset to zero

https://user-images.githubusercontent.com/127308740/225426936-562e20cb-b2b9-432f-8944-0d0431da508f.mp4


https://user-images.githubusercontent.com/127308740/225426946-18ab47a1-6ed1-4551-bb4f-8924692825db.mp4

**Comment by jfedor2** [comment]
Different behavior is possibly caused by the fact that the axes are relative in the Pro's report descriptor. While as I said I think this is incorrect, it also means that it doesn't have to send zeros when the movement stops, just like a normal 2D mouse doesn't.

You can dump the SpaceExplorer's report descriptor to see if the usages are absolute in there.

As for the buttons causing movement, I don't have any new ideas, you can try connecting the ~~Pro~~ Classic directly to the PC and capturing traffic with Wireshark while pressing buttons and not moving the joystick. Then we can try to see if it looks like random noise or some protocol misinterpretation.

The 0xFF00... usages are vendor-specific, who knows what they're for.

**Comment by Platidors** [comment]
I'll try when I get home. Could you help me to set the deadzone through the Expression settings. At least approximate values and code format?

**Comment by Platidors** [comment]
> You can dump the SpaceExplorer's report descriptor to see if the usages are absolute in there.
[SpaceExplorer.zip](https://github.com/jfedor2/magellan-spacemouse/files/10993179/SpaceExplorer.zip)

**Comment by Platidors** [comment]
> As for the buttons causing movement, I don't have any new ideas, you can try connecting the ~Pro~ Classic directly to the PC and capturing traffic with Wireshark while pressing buttons and not moving the joystick. Then we can try to see if it looks like random noise or some protocol misinterpretation.
[SpaceMouseClassicKey.zip](https://github.com/jfedor2/magellan-spacemouse/files/10993387/SpaceMouseClassicKey.zip)
Pressed 1 2 3 4 5 6 7 8 9 *

**Comment by jfedor2** [comment]
Well, it really is sending those axis inputs, I don't see any "logical" pattern there or anything that would suggest it's the remapper's fault.

To configure dead zones, first set up six expressions, one for each axis.

You have to use the following hex codes:

0x00010030: X
0x00010031: Y
0x00010032: Z
0x00010033: RX
0x00010034: RY
0x00010035: RZ

Each expression will look the same, only the hex code will be different. Of course you can also set the dead zone per-axis if you want.

For example let's say Expression 1 will be for axis X:

```
0x00010030 input_state dup abs 10 gt mul scaling mul
```

It's explained in detail here:

https://github.com/jfedor2/hid-remapper/blob/master/EXPRESSIONS.md

But basically we're fetching the axis state, duplicating it, taking its absolute value, checking if it's greater than 10 or not (this is the dead zone), which gives either a 0 or a 1, multiplying it by the original value, leaving us with either zero if it was inside the dead zone or the original value if it was not. At the end we fetch the scaling value from the mapping and multiply by it. That way you can set the axis inversions as before.

Assuming you configured Expressions 1-6 this way:

![Screenshot from 2023-03-17 00-01-06](https://user-images.githubusercontent.com/2271721/225771360-5ee4125a-d0b5-49ff-bed2-774f2d76e352.png)

You can then configure the mappings like so:

![Screenshot from 2023-03-16 23-56-50](https://user-images.githubusercontent.com/2271721/225771095-c3a77413-786a-482c-a7c5-2654cfc2c31d.png)

And in theory it should work.

**Comment by jfedor2** [comment]
Here's a JSON with the above configuration.

[spacemouse-deadzones.zip](https://github.com/jfedor2/magellan-spacemouse/files/10996497/spacemouse-deadzones.zip)

**Comment by jfedor2** [comment]
It would also be possible to program it to ignore axis inputs when a button is pressed.

**Comment by Platidors** [comment]
The filter seems to be working well. Pressing the buttons does not lay down the axes. It is enough to filter the axes RX, RY, RZ. Maybe there is a way to return the movement data to 0 when there is no action on the joystick?

**Comment by jfedor2** [comment]
Is this a problem in practice? In real software?

**Comment by Platidors** [comment]
It is not possible to move the center of rotation of an object in solidworks.  The space mouse cursor does not move over the part, but is always in the center

**Comment by jfedor2** [comment]
I'm not familiar with Solidworks. Is that related? How did you expect it to work?

**Comment by Platidors** [comment]
When the coordinates of all axes are set to 0, the space mouse cursor is released and moved to the part in the center of the screen on another part of the part. This allows it to rotate around the new center. You seem to intercept the part with your hand when you move it.
  I have interesting news. I tried connecting the space guide to the HID-remapper and it worked. The HID-remapper turns out to be a universal device. In the 3DxTest program, the coordinates are also not returned to 0.

**Comment by jfedor2** [comment]
Oh, I see!

Yeah, it should be possible to make it send zeros, let me think how to do it best.

And yeah, HID Remapper was originally built to map anything to anything. :) The original one emulates mouse and keyboard so the output needed to be modified, but on the input side we pretty much support anything. You could connect a flight sim joystick or a game pad and use it as a 3D mouse, assuming you set up the mappings in some reasonable way.

(Well, that's not entirely true, the current version limits the report size to 14 bytes, but that's just a quick hack to handle the SpaceMouse Classic's reports being split into two packets, it could be fixed properly to work with any report size.)

**Comment by jfedor2** [comment]
This should send zeros after the movement stops and hopefully not break anything else:

[remapper_spacemouse10.zip](https://github.com/jfedor2/magellan-spacemouse/files/11009645/remapper_spacemouse10.zip)

(The B side didn't change, I just included it for completeness.)

**Comment by Platidors** [comment]
It works, but not always. RX, RY, RZ get up to 0 always, TX, TY, TZ get up to 0 sometimes. In solidworks cursor moves but not every time you release the joystick

https://user-images.githubusercontent.com/127308740/226142321-cc9bc208-fa1b-47d1-9cf3-9c564fc6a278.mp4

dead zones included on all axes

**Comment by jfedor2** [comment]
Interesting.

I took a closer look at what my SpaceMouse Compact sends and noticed two things. First, even though the translations and rotations are separate report IDs, it always sends them in pairs. Second, when the movement stops, it actually sends the all-zero reports three times.

I should probably try to emulate at least the first thing.

**Comment by Platidors** [comment]
Perhaps in this way 3dconnexion tried to protect its manipulators from copying

**Comment by jfedor2** [comment]
OK, this should always send the axes reports in pairs.

[remapper_spacemouse11.zip](https://github.com/jfedor2/magellan-spacemouse/files/11012513/remapper_spacemouse11.zip)

**Comment by Platidors** [comment]
On this firmware, the space mouse works great. In all programs, the movements became smooth. In the 3DxTest program, all coordinates go to 0 after releasing the joystick. In the solidworks program, the cursor moves well. I still use your filters to set dead zones for all axes. And it works fine. Without them, false positives of the joystick appear when a key is pressed. This is apparently the specifics of the old matipulators. I tried to connect a modern space explorer. On this firmware, the joystick is not processed very well, intermittent movements are noticeable. So this firmware is not universal for all space mouse. Video when connecting the space explorer.


https://user-images.githubusercontent.com/127308740/226259257-efa7f1dd-0aa4-42f0-9da3-930747c345f7.mp4

**Comment by jfedor2** [comment]
Yeah, this is probably because we're now sending two reports for every received motion report and the Explorer, unlike the Classic, sends the translations and rotations in separate reports. So for the Explorer, every time we receive a X,Y,Z report, we send a (normal) X,Y,Z report and a (zero) RX,RY,RZ report and for each RX,RY,RZ report we send a (zero) X,Y,Z report and a (normal) RX,RY,RZ report.

My goal was to make it universal and a minimal modification of the regular HID Remapper, but it turned out a few device-specific hacks were necessary. Maybe I will find a way to make it more generic in the future.

**Comment by Platidors** [comment]
In the rs-232 space mouse Maggelan version of the emulator, there are no such problems as double reporting, turning the joystick when pressing the buttons, not returning the coordinates to 0?

**Comment by jfedor2** [comment]
The Magellan adapter doesn't use HID Remapper as a base, it's much simpler code that just happened to avoid those issues, even though I wasn't aware of them at the time.
